examples/cluster/plot_optics.py

```diff
@@ -0,0 +1,93 @@
+"""
+===================================
+Demo of OPTICS clustering algorithm
+===================================
+
+Finds core samples of high density and expands clusters from them.
+This example uses data that is generated so that the clusters have
+different densities.
+"""
+
+# Authors: Shane Grigsby <refuge@rocktalus.com>
+#          Amy X. Zhang <axz@mit.edu>
+# License: BSD 3 clause
+
+
+from sklearn.cluster import OPTICS
+import matplotlib.gridspec as gridspec
+
+
+import numpy as np
+
+import matplotlib.pyplot as plt
+
+# Generate sample data
+
+np.random.seed(0)
+n_points_per_cluster = 250
+
+C1 = [-5, -2] + .8 * np.random.randn(n_points_per_cluster, 2)
+C2 = [4, -1] + .1 * np.random.randn(n_points_per_cluster, 2)
+C3 = [1, -2] + .2 * np.random.randn(n_points_per_cluster, 2)
+C4 = [-2, 3] + .3 * np.random.randn(n_points_per_cluster, 2)
+C5 = [3, -2] + 1.6 * np.random.randn(n_points_per_cluster, 2)
+C6 = [5, 6] + 2 * np.random.randn(n_points_per_cluster, 2)
+X = np.vstack((C1, C2, C3, C4, C5, C6))
+
+clust = OPTICS(min_samples=9, rejection_ratio=0.5)
+
+# Run the fit
+clust.fit(X)
+
+_, labels_025 = clust.extract_dbscan(0.25)
+_, labels_075 = clust.extract_dbscan(0.75)
+
+space = np.arange(len(X))
+reachability = clust.reachability_[clust.ordering_]
+labels = clust.labels_[clust.ordering_]
+
+plt.figure(figsize=(10, 7))
+G = gridspec.GridSpec(2, 3)
+ax1 = plt.subplot(G[0, :])
+ax1.set_ylabel('Reachability (epsilon distance)')
+ax1.set_title('Reachability Plot')
+ax2 = plt.subplot(G[1, 0])
+ax2.set_title('Automatic Clustering')
+ax3 = plt.subplot(G[1, 1])
+ax3.set_title('Clustering at 0.25 epsilon cut')
+ax4 = plt.subplot(G[1, 2])
+ax4.set_title('Clustering at 0.75 epsilon cut')
+
+# Reachability plot
+color = ['g.', 'r.', 'b.', 'y.', 'c.']
+for k, c in zip(range(0, 5), color):
+    Xk = space[labels == k]
+    Rk = reachability[labels == k]
+    ax1.plot(Xk, Rk, c, alpha=0.3)
+ax1.plot(space[labels == -1], reachability[labels == -1], 'k.', alpha=0.3)
+ax1.plot(space, np.ones_like(space) * 0.75, 'k-', alpha=0.5)
+ax1.plot(space, np.ones_like(space) * 0.25, 'k-.', alpha=0.5)
+
+# OPTICS
+color = ['g.', 'r.', 'b.', 'y.', 'c.']
+for k, c in zip(range(0, 5), color):
+    Xk = X[clust.labels_ == k]
+    ax2.plot(Xk[:, 0], Xk[:, 1], c, alpha=0.3)
+ax2.plot(X[clust.labels_ == -1, 0], X[clust.labels_ == -1, 1], 'k+', alpha=0.1)
+
+# DBSCAN at 0.25
+color = ['g.', 'greenyellow.', 'olive.', 'r.', 'b.', 'c.']
+for k, c in zip(range(0, 6), color):
```

Comment:
But now it fails as `labels_075` and `labels_025` start at 1 and not at 0.
To avoid problems here, maybe use `np.unique(labels)[1:]` and `np.unique(labels_075)[1:]`.

But more generally, why does `extract_dbscan` uses labelling from 1 and `_extract_optics` from 0?
What does `DBSCAN` use?
We should really use consistent indexing !
And we should add a test to enforce this consistency.

Reply:
Thanks for tracking this-- fixed now. 

There is already a test for this (sorta)-- https://github.com/scikit-learn/scikit-learn/pull/10015

That test passes on the previous build because it just checks for if labeling starts at 0 or -1... in the case where -1 is present (noise), it doesn't care where actual labeling of the **clusters** starts.

I can write a test for this, where it verifies that clusters are sequentially labeled starting from either -1 or 0. That would also catch any cases where labeling is monotonic, but not continuous (i.e., labels like [-1, 0, 1, 2, 4, 5]). Since it's pretty common to take the max of labels_, this seems like a good idea in general.

With that in mind, it might make sense to add the test to clustering estimators in general. I think this should be enforced at the clustering estimator level, rather than at the individual algorithm...